aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs

```diff
@@ -0,0 +1,196 @@
+        })
+    }
+
+    /// Run the processor and pass user-defined validation logic
+    pub async fn run<F>(
+        &self,
+        processor: &impl ProcessorTrait,
+        txn_verion: u64,
+        generate_files: bool, // flag to control file generation
+        output_path: Option<String>, // Optional custom output path
+        verification_f: F,
+    ) -> anyhow::Result<serde_json::Value>
+    where
+        F: FnOnce(&str) -> anyhow::Result<serde_json::Value> + Send + Sync + 'static,
+    {
+        let transactions = self.transaction_batches.clone();
+        let transactions_response = vec![TransactionsResponse {
+            transactions,
+            ..TransactionsResponse::default()
+        }];
+
+        self.setup_mock_grpc(transactions_response, 1).await;
+
+        // Run the processor
+        processor.run_processor().await?;
+
+        // Pass the DB URL for user-defined validation
+        let db_url = self.database.get_db_url();
+        let db_values = verification_f(&db_url)?;
+
+        // Conditionally generate output files if the `generate_files` flag is true
+        if generate_files {
+            // Use the provided output path or fall back to default
+            let processor_name = processor.name();
+            let processor_name = "events_processor"; //  TODO: add a function to get name in runtime
+            let txn_version = txn_verion.to_string();
+            let output_dir = output_path.unwrap_or_else(|| {
+                format!(
+                    "expected_db_output_files/imported_testnet_txns/{}/{}_{}.json",
+                    processor_name, processor_name, txn_version
+                )
+            });
+
+            // Extract the directory path from the full file path
+            if let Some(parent_dir) = Path::new(&output_dir).parent() {
+                // Create the directory if it doesn't exist
+                fs::create_dir_all(parent_dir)?;
+            }
+
+            // Save the database output as JSON to the specified path
+            fs::write(output_dir.clone(), serde_json::to_string_pretty(&db_values)?)?;
+            println!("[INFO] Generated output file at: {}", output_dir);
+        } else {
+            println!("[INFO] Skipping file generation as requested.");
+        }
+
+        Ok(db_values)
+    }
+
+    /// Helper function to set up and run the mock GRPC server.
+    async fn setup_mock_grpc(&self, transactions: Vec<TransactionsResponse>, chain_id: u64) {
+        println!("received transactions size: {:?}", transactions.len());
+        let mock_grpc_server = MockGrpcServer {
+            transactions,
+            chain_id,
+        };
+
+        // Start the Mock GRPC server
+        tokio::spawn(async move {
+            println!("Starting Mock GRPC server");
+            mock_grpc_server.run().await;
+        });
+    }
+
+    pub fn create_transaction_stream_config(&self,
+        starting_version: u64,
+        ending_version: u64,
```

Comment:
This could be cleaned up a bit imo, maybe this should accept a vec of transactions, and then dynamically sort them by version and get the starting/ending version from the vec?